enhancements/network/routed-ingress-support-for-primary-udn-attached-vms-with-static-ips.md

```diff
@@ -0,0 +1,845 @@
+    message: "Duplicate MAC 00:1a:4b:12:34:56 found in entries"
+    lastTransitionTime: "2025-05-13T12:05:00Z"
+
+  - type: DuplicateMACAddresses
+    status: "True"
+    reason: DuplicateMACFound
+    message: "MAC address 00:1a:4b:12:34:56 appears 2 times"
+    lastTransitionTime: "2025-05-13T12:05:00Z"
+
+  - type: DuplicateIPAddresses
+    status: "False"
+    reason: NoIPDuplicates
+    message: "All IP addresses are unique per MAC"
+    lastTransitionTime: "2025-05-13T12:05:00Z"
+```
+
+On a follow-up, we could reconcile the `IPPool` CR with the MAC <-> IPs
+association of all workloads attached to that network - i.e. also for the
+workloads being created without MAC / IP requests. That would be useful to the
+network admin / service engineer to get the full view on whatever's connected
+to a network at any given time. This would require a separate enhancement to
+get better understanding of the use-cases, and how to implement said feature.
+
+#### De-centralized IP management
+
+This approach requires having N CRs with a 1:1 association between a primary
+UDN attachment and the MAC and IPs it had on the original platform.
+
+We could either introduce a new CRD with IPs and MAC association (which would
+deprecate the IPAMClaim CRD), or, assuming clunkiness in all its glory, we
+could change the IPAMClaim CRD to have MAC and IP addresses being requested in
+the spec stanza as well.
+
+OVN-Kubernetes would read the CR, attempt to reserve the requested MAC and IP,
+and then persist that information in the IPAMClaim status - reporting a
+successful sync in the `IPAMClaim` status - or a failure otherwise.
+
+In this option, the [CNV to OVN-Kubernetes API](#cnv-ovnk-api) **remains** as
+defined today - i.e. OVN-Kubernetes will be pointed to an `IPAMClaim` via a
+particular annotation on the pod.
+
+The flow is described in more detail in the
+[De-centralized IP management workflow](#de-centralized-ip-management-workflow) section.
+
+##### De-centralized IP management workflow
+
+The [de-centralized IP management](#de-centralized-ip-management) flow is
+described in the following sequence diagram:
+
+```mermaid
+sequenceDiagram
+actor Admin
+actor Project admin
+
+participant MTV
+participant CNV
+participant o as OVN-Kubernetes
+
+Admin ->> o: cluster UDN
+o -->> Admin: OK
+
+Admin ->> CNV: provision IPAMClaim w/ MAC and IP requests
+CNV -->> Admin: OK
+
+Project admin ->> MTV: import VM
+MTV ->> CNV: create VM(name=<...>, primaryUDNMac=origMAC)
+CNV ->> CNV: ips = getIPsForMAC(mac=origMAC)
+CNV ->> o: create pod(mac=origMAC, IPs=ips)
+o -->> CNV: OK
+CNV -->> MTV: OK
+MTV -->> Project admin: OK
+```
+
+Hence, the required changes would be:
+- the `IPAMClaim` CRD status sub-resource will need to be updated, adding a
+  `Conditions` array. This way we will be able to report errors back to the
+  user (e.g. the desired IP allocation is already in use).
+- the `IPAMClaim` CRD spec will need to be updated, adding `MAC` and `IPs`
+  requests.
```

Comment:
With the planed change to extend the NSE annotation to ref IPAMClaim CR, what would take precedence the NSE IP/MAC request attributes or the IPAMClain CR spec?
What if both are specified (NSE IP/MAC request and IPAMClaim spec IP/MAC)?
If its not the right place to discuss please let me know and  feel free to resolve this thread.

Reply:
No, this question is relevant.

I wouldn't add any validation at this first stage, but, to be honest, if I had to choose a course of action, I'd error out and prevent creating the VM if the NSE has different IPs than the IPAMClaim status (when they're present - keep in mind that initially, the ipam claim will exist, but won't have any IPs in the status).